File: scripts/backfill/normalizers.py
"""
String -> enum normalizers driven by shared/enums.json. These depend only on
the raw strings UFC Stats prints, not on page structure, so they are written
and tested now. Every unknown value raises SchemaAssertionError.

Mirror: workers/ufc-stats-ingest/src/normalizers.mjs (same enums.json, same
fixtures in shared/tests/normalizer_fixtures.json).
"""
from __future__ import annotations

import re
from typing import Optional

from common import ENUMS, SchemaAssertionError

_RND = re.compile(r"^\s*(\d+)\s*Rnd", re.I)
_MMSS = re.compile(r"^\s*(\d+):(\d{1,2})\s*$")
_X_OF_Y = re.compile(r"^\s*(\d+)\s+of\s+(\d+)\s*$", re.I)
_HEIGHT = re.compile(r"^\s*(\d+)'\s*(\d+)\"?\s*$")
_RECORD = re.compile(r"Record:\s*(\d+)-(\d+)-(\d+)(?:\s*\((\d+)\s*NC\))?", re.I)


def norm_method(raw: str, url: str) -> str:
    """'Decision - Unanimous' -> DEC_U. First line of a two-line method cell only."""
    key = (raw or "").strip().splitlines()[0].strip() if raw and raw.strip() else ""
    m = ENUMS["method"]["map"]
    if key in m:
        return m[key]
    raise SchemaAssertionError(url, f"unknown method {raw!r}")


# "TUF <edition> <side A> vs. <side B> " where a division follows. Anchored at
# the start and requiring the TUF prefix, so it matches the international
# seasons' verified label format and nothing else. Only the formats we have
# actually seen in source data belong here; more are added when a fixture
# proves them, not in anticipation.
_TUF_MATCHUP_PREFIX = re.compile(
    r"^\s*TUF\s+"
    r"[A-Za-z][A-Za-z.'\-]*(?:\s+[A-Za-z][A-Za-z.'\-]*)*?"
    r"\s+vs\.?\s+"
    r"[A-Za-z][A-Za-z.'\-]*(?:\s+[A-Za-z][A-Za-z.'\-]*)*?"
    r"\s+(?=\S)",
    re.I,
)

# The tournament era, 1993-1999. Four windows failed on it — B-1995, B-1996,
# B-1997 and B-1999 — and the four events they died on say what the two
# leftovers are:
#
#   Ultimate Ultimate '95, Ultimate Ultimate '96   the EVENT SERIES name lands
#   Ultimate Japan, Ultimate Japan 2               in the division slot, the
#                                                  same way Brazil and China
#                                                  already do above.
#
#   "Tournament Semifinal Bout", "Alternate Bout"  the BRACKET POSITION lands
#                                                  there too. Where a fighter
#                                                  stood in a tournament is not
#                                                  what weight he stood at.
#
# Both lists are closed: only strings our own failing events produced. A new
# series or a new round word is added when a window proves it, not in
# anticipation of one.
_EARLY_SERIES = re.compile(r"\b(Ultimate Ultimate|Ultimate Japan|Japan)\b", re.I)
_BRACKET_ROUND = re.compile(r"\b(Semifinals?|Quarterfinals?|Alternate|Final)\b", re.I)

# What is left once a series name goes. The real label is
#   "Ultimate Ultimate '96 Tournament Title Bout"
# and the generic numeric strip above has already taken the 96, so removing the
# series leaves a lone apostrophe. It is punctuation from a year that no longer
# exists, not a division, and without dropping it the rescue sees a non-empty
# remainder and refuses a label it has in fact fully resolved. Confined to this
# rescue, where the result must still match a known division or be empty, so it
# cannot make anything else resolve.
_ORPHAN_PUNCT = re.compile(r"[‘’'`.,\-]+")


def norm_weight_class(raw: str, url: str) -> dict:
    """'UFC Women's Bantamweight Title Bout' -> {weight_class: BANTAMWEIGHT, is_womens: True, is_title: True}.
    Empty string (some very old bouts) -> weight_class None."""
    wc = ENUMS["weight_class"]
    s = (raw or "").strip()
    if not s:
        return {"weight_class": None, "is_womens": False, "is_title": False}
    is_womens = wc["womens_marker"].lower() in s.lower()
    is_title = any(t.lower() in s.lower() for t in wc["title_markers"])
    core = s
    # Feeder-series prefix. "Road to UFC 3 Bantamweight Tournament Title Bout"
    # is an ordinary Bantamweight title bout staged inside a regional series,
    # and the series name is not part of the division. It has to go before the
    # generic token pass, which would otherwise remove "UFC" and the season
    # number and leave the unmatchable remainder "Road to Bantamweight".
    # The raw label is still stored verbatim in weight_class_raw, so the
    # tournament context is preserved rather than discarded.
    core = re.sub(r"^\s*Road to UFC\b\s*\d*", " ", core, flags=re.I)
    core = re.sub(re.escape(wc["womens_marker"]), " ", core, flags=re.I)
    for tok in wc["strip_tokens"]:
        core = re.sub(r"\b" + re.escape(tok) + r"\b", " ", core, flags=re.I)
    core = re.sub(r"\b\d+\b", " ", core)                      # "Ultimate Fighter 1 ..."
    core = re.sub(r"\b(Latin America|Brazil|China|Nations|Australia vs\.? UK|Team [A-Za-z]+)\b", " ", core, flags=re.I)
    core = re.sub(r"\s+", " ", core).strip()
    if not core:   # e.g. "UFC 2 Tournament Title Bout": early tournaments carried no weight class
        return {"weight_class": None, "is_womens": is_womens, "is_title": is_title}
    for k, v in wc["map"].items():
        if core.lower() == k.lower():
            return {"weight_class": v, "is_womens": is_womens, "is_title": is_title}

    # Last chance, and only for a TUF label whose matchup leaked into the
    # division. B-2014 stopped on
    #   "TUF Nations Canada vs. Australia Middleweight Tournament Title Bout"
    # which reduces to "TUF Canada vs. Australia Middleweight": the country
    # pairing is part of the SEASON's name, not of the weight class.
    #
    # Deliberately a rescue rather than another pass over every label. It runs
    # only after normal resolution has already failed, and it is accepted only
    # if what remains is a division we recognise, so it cannot change the
    # outcome of any label that resolves today and cannot turn an unknown
    # division into a known one. The anchor is the literal "TUF" prefix plus an
    # "X vs. Y" pairing followed by something else; a label that merely
    # contains a country name is untouched and still fails closed.
    rescued = _TUF_MATCHUP_PREFIX.sub(" ", core, count=1).strip()
    if rescued and rescued != core:
        for k, v in wc["map"].items():
            if rescued.lower() == k.lower():
                return {"weight_class": v, "is_womens": is_womens, "is_title": is_title}

    # Second rescue, on the same terms as the first: it runs only after normal
    # resolution has failed, it strips only the closed lists above, and it is
    # accepted only if what remains is a division we recognise or nothing at
    # all. So it cannot change any label that resolves today, and it cannot
    # invent a division — "Japan Featherweight" is still not a weight class and
    # still fails closed.
    #
    # Resolving to nothing is a real answer here, not a failure to find one.
    # "Alternate Bout" and "UFC Japan Tournament Title Bout" carry no division
    # because those bouts had none, which is exactly what the empty-core branch
    # above already concludes for "UFC 2 Tournament Title Bout".
    early = _BRACKET_ROUND.sub(" ", _EARLY_SERIES.sub(" ", core))
    early = _ORPHAN_PUNCT.sub(" ", early)
    early = re.sub(r"\s+", " ", early).strip()
    if early != core:
        if not early:
            return {"weight_class": None, "is_womens": is_womens, "is_title": is_title}
        for k, v in wc["map"].items():
            if early.lower() == k.lower():
                return {"weight_class": v, "is_womens": is_womens, "is_title": is_title}

    raise SchemaAssertionError(url, f"unknown weight class {raw!r} (core={core!r})")


def norm_stance(raw: Optional[str], url: str) -> Optional[str]:
    s = (raw or "").strip()
    st = ENUMS["stance"]
    if s in st["null_values"]:
        return None
    if s in st["map"]:
        return st["map"][s]
    raise SchemaAssertionError(url, f"unknown stance {raw!r}")


def scheduled_rounds(time_format: Optional[str], url: str) -> Optional[int]:
    s = (time_format or "").strip()
    sr = ENUMS["scheduled_rounds"]
    if any(s.startswith(nv) for nv in sr["null_values"] if nv) or s == "":
        return None
    m = _RND.match(s)
    if m:
        return int(m.group(1))
    raise SchemaAssertionError(url, f"unknown time format {time_format!r}")


def mmss_to_sec(raw: Optional[str], url: str) -> Optional[int]:
    s = (raw or "").strip()
    if s in ("", "--"):
        return None
    m = _MMSS.match(s)
    if not m:
        raise SchemaAssertionError(url, f"bad mm:ss {raw!r}")
    return int(m.group(1)) * 60 + int(m.group(2))


def x_of_y(raw: Optional[str], url: str) -> tuple[Optional[int], Optional[int]]:
    s = (raw or "").strip()
    if s in ("", "--", "---"):
        return None, None
    m = _X_OF_Y.match(s)
    if not m:
        raise SchemaAssertionError(url, f"bad 'x of y' {raw!r}")
    return int(m.group(1)), int(m.group(2))


def int_or_none(raw: Optional[str], url: str) -> Optional[int]:
    s = (raw or "").strip()
    if s in ("", "--", "---"):
        return None
    if not s.isdigit():
        raise SchemaAssertionError(url, f"bad int {raw!r}")
    return int(s)


def pct_or_none(raw: Optional[str], url: str) -> Optional[float]:
    s = (raw or "").strip().rstrip("%")
    if s in ("", "--"):
        return None
    try:
        return float(s)
    except ValueError:
        raise SchemaAssertionError(url, f"bad percent {raw!r}")


def num_or_none(raw: Optional[str], url: str) -> Optional[float]:
    s = (raw or "").strip()
    if s in ("", "--"):
        return None
    try:
        return float(s)
    except ValueError:
        raise SchemaAssertionError(url, f"bad number {raw!r}")


def height_in(raw: Optional[str], url: str) -> Optional[float]:
    s = (raw or "").strip()
    if s in ("", "--"):
        return None
    m = _HEIGHT.match(s)
    if not m:
        raise SchemaAssertionError(url, f"bad height {raw!r}")
    return int(m.group(1)) * 12 + int(m.group(2))


def reach_in(raw: Optional[str], url: str) -> Optional[float]:
    s = (raw or "").strip().rstrip('"').strip()
    return num_or_none(s, url)


def weight_lbs(raw: Optional[str], url: str) -> Optional[float]:
    s = re.sub(r"\s*lbs\.?$", "", (raw or "").strip(), flags=re.I)
    return num_or_none(s, url)


def record(raw: str, url: str) -> dict:
    """'Record: 20-3-0 (1 NC)' -> {record_w:20, record_l:3, record_d:0, record_nc:1}"""
    m = _RECORD.search(raw or "")
    if not m:
        raise SchemaAssertionError(url, f"bad record {raw!r}")
    return {"record_w": int(m.group(1)), "record_l": int(m.group(2)), "record_d": int(m.group(3)),
            "record_nc": int(m.group(4)) if m.group(4) else 0}


def event_date(raw: str, url: str) -> str:
    """'February 15, 2026' -> '2026-02-15'"""
    import datetime as dt
    s = (raw or "").strip()
    for fmt in ("%B %d, %Y", "%b %d, %Y"):
        try:
            return dt.datetime.strptime(s, fmt).date().isoformat()
        except ValueError:
            pass
    raise SchemaAssertionError(url, f"bad event date {raw!r}")


def dob(raw: Optional[str], url: str) -> Optional[str]:
    s = (raw or "").strip()
    if s in ("", "--"):
        return None
    return event_date(s, url)


def scorecards(details: str) -> Optional[list[dict]]:
    """'Sal D'Amato 29-28. Derek Cleary 29-28. Chris Lee 28-29.' -> [{judge, score}, ...]; None if no scores."""
    found = re.findall(r"([A-Za-z][A-Za-z .'\-]+?)\s+(\d{1,3}\s*-\s*\d{1,3})\s*\.?", details or "")
    if not found:
        return None
    return [{"judge": j.strip(), "score": re.sub(r"\s+", "", sc)} for j, sc in found]

File: workers/ufc-stats-ingest/src/normalizers.mjs
/* String -> enum normalizers, Worker side. Mirror of scripts/backfill/normalizers.py.
 * Driven by shared/enums.json. In the repo that is ../../../shared/enums.json;
 * scripts/package_worker.py copies shared/ into src/shared/ for the deployable
 * zip so C:\Workers\ufc-stats-ingest is self-contained. Every unknown value
 * throws SchemaAssertionError.
 */
import ENUMS from './shared/enums.json' with { type: 'json' };
import { SchemaAssertionError } from './ufcstats.mjs';

const esc = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

export function normMethod(raw, url) {
  const key = String(raw || '').trim().split(/\r?\n/)[0].trim();
  const m = ENUMS.method.map;
  if (Object.prototype.hasOwnProperty.call(m, key)) return m[key];
  /* ESPN prints some cards (Road to UFC) with the finish detail in parentheses:
   * "Submission (Kimura)", "TKO (Knee)", "Decision - Split (Draw)". The base
   * must itself be a known method; the detail never invents one. A split draw
   * is the only parenthetical that changes the outcome. */
  const paren = /^(.+?)\s*\(([^()]+)\)$/.exec(key);
  if (paren) {
    const base = paren[1].trim(), detail = paren[2].trim();
    if (/^draw$/i.test(detail) && /^decision - (split|majority|unanimous)$/i.test(base)) return m['Draw'];
    if (/^(technical submission|submission|ko\/tko|tko|ko)$/i.test(base) && Object.prototype.hasOwnProperty.call(m, base) && !/draw|no contest/i.test(detail)) return m[base];
  }
  throw new SchemaAssertionError(url, `unknown method ${JSON.stringify(raw)}`);
}

export function normWeightClass(raw, url) {
  const wc = ENUMS.weight_class;
  const s = String(raw || '').trim();
  if (!s) return { weight_class: null, is_womens: false, is_title: false };
  const lower = s.toLowerCase();
  const is_womens = lower.includes(wc.womens_marker.toLowerCase());
  const is_title = wc.title_markers.some((t) => lower.includes(t.toLowerCase()));
  /* Feeder-series prefix. "Road to UFC 3 Bantamweight Tournament Title Bout"
   * is an ordinary Bantamweight title bout staged inside a regional series,
   * and the series name is not part of the division. It has to go before the
   * generic token pass, which would otherwise remove "UFC" and the season
   * number and leave the unmatchable remainder "Road to Bantamweight".
   * Kept identical to scripts/backfill/normalizers.py: the scheduled worker
   * and the historical backfill must agree, or the same card parses one way
   * live and another way on replay. */
  let core = s.replace(/^\s*Road to UFC\b\s*\d*/i, ' ');
  core = core.replace(new RegExp(esc(wc.womens_marker), 'ig'), ' ');
  for (const tok of wc.strip_tokens) core = core.replace(new RegExp(`\\b${esc(tok)}\\b`, 'ig'), ' ');
  core = core.replace(/\b\d+\b/g, ' ');
  core = core.replace(/\b(Latin America|Brazil|China|Nations|Australia vs\.? UK|Team [A-Za-z]+)\b/ig, ' ');
  core = core.replace(/\s+/g, ' ').trim();
  if (!core) return { weight_class: null, is_womens, is_title };   /* "UFC 2 Tournament Title Bout" */
  for (const [k, v] of Object.entries(wc.map)) {
    if (core.toLowerCase() === k.toLowerCase()) return { weight_class: v, is_womens, is_title };
  }
  throw new SchemaAssertionError(url, `unknown weight class ${JSON.stringify(raw)} (core=${JSON.stringify(core)})`);
}

export function normStance(raw, url) {
  const s = String(raw || '').trim();
  const st = ENUMS.stance;
  if (st.null_values.includes(s)) return null;
  if (Object.prototype.hasOwnProperty.call(st.map, s)) return st.map[s];
  throw new SchemaAssertionError(url, `unknown stance ${JSON.stringify(raw)}`);
}

export function scheduledRounds(timeFormat, url) {
  const s = String(timeFormat || '').trim();
  const sr = ENUMS.scheduled_rounds;
  if (s === '' || sr.null_values.some((nv) => nv && s.startsWith(nv))) return null;
  const m = s.match(/^\s*(\d+)\s*Rnd/i);
  if (m) return Number(m[1]);
  throw new SchemaAssertionError(url, `unknown time format ${JSON.stringify(timeFormat)}`);
}

export function mmssToSec(raw, url) {
  const s = String(raw || '').trim();
  if (s === '' || s === '--') return null;
  const m = s.match(/^(\d+):(\d{1,2})$/);
  if (!m) throw new SchemaAssertionError(url, `bad mm:ss ${JSON.stringify(raw)}`);
  return Number(m[1]) * 60 + Number(m[2]);
}

export function xOfY(raw, url) {
  const s = String(raw || '').trim();
  if (['', '--', '---'].includes(s)) return [null, null];
  const m = s.match(/^(\d+)\s+of\s+(\d+)$/i);
  if (!m) throw new SchemaAssertionError(url, `bad 'x of y' ${JSON.stringify(raw)}`);
  return [Number(m[1]), Number(m[2])];
}

export function intOrNull(raw, url) {
  const s = String(raw || '').trim();
  if (['', '--', '---'].includes(s)) return null;
  if (!/^\d+$/.test(s)) throw new SchemaAssertionError(url, `bad int ${JSON.stringify(raw)}`);
  return Number(s);
}

export function numOrNull(raw, url) {
  const s = String(raw || '').trim();
  if (s === '' || s === '--') return null;
  const n = Number(s);
  if (Number.isNaN(n)) throw new SchemaAssertionError(url, `bad number ${JSON.stringify(raw)}`);
  return n;
}

export function pctOrNull(raw, url) {
  return numOrNull(String(raw || '').trim().replace(/%$/, ''), url);
}

export function heightIn(raw, url) {
  const s = String(raw || '').trim();
  if (s === '' || s === '--') return null;
  const m = s.match(/^(\d+)'\s*(\d+)"?$/);
  if (!m) throw new SchemaAssertionError(url, `bad height ${JSON.stringify(raw)}`);
  return Number(m[1]) * 12 + Number(m[2]);
}

export function reachIn(raw, url) {
  return numOrNull(String(raw || '').trim().replace(/"$/, '').trim(), url);
}

export function weightLbs(raw, url) {
  return numOrNull(String(raw || '').trim().replace(/\s*lbs\.?$/i, ''), url);
}

export function record(raw, url) {
  const m = String(raw || '').match(/Record:\s*(\d+)-(\d+)-(\d+)(?:\s*\((\d+)\s*NC\))?/i);
  if (!m) throw new SchemaAssertionError(url, `bad record ${JSON.stringify(raw)}`);
  return { record_w: Number(m[1]), record_l: Number(m[2]), record_d: Number(m[3]), record_nc: m[4] ? Number(m[4]) : 0 };
}

const MONTHS = { january: 1, february: 2, march: 3, april: 4, may: 5, june: 6, july: 7, august: 8, september: 9, october: 10, november: 11, december: 12,
  jan: 1, feb: 2, mar: 3, apr: 4, jun: 6, jul: 7, aug: 8, sep: 9, sept: 9, oct: 10, nov: 11, dec: 12 };

export function eventDate(raw, url) {
  const m = String(raw || '').trim().match(/^([A-Za-z]+)\.?\s+(\d{1,2}),\s*(\d{4})$/);
  const mon = m && MONTHS[m[1].toLowerCase()];
  if (!mon) throw new SchemaAssertionError(url, `bad event date ${JSON.stringify(raw)}`);
  return `${m[3]}-${String(mon).padStart(2, '0')}-${String(m[2]).padStart(2, '0')}`;
}

export function dob(raw, url) {
  const s = String(raw || '').trim();
  if (s === '' || s === '--') return null;
  return eventDate(s, url);
}

export function scorecards(details) {
  const found = [...String(details || '').matchAll(/([A-Za-z][A-Za-z .'\-]+?)\s+(\d{1,3}\s*-\s*\d{1,3})\s*\.?/g)];
  if (!found.length) return null;
  return found.map((m) => ({ judge: m[1].trim(), score: m[2].replace(/\s+/g, '') }));
}

/* Scheduled rounds for an ESPN competition (2026-09-24 audit). ESPN's format.regulation.periods is NOT the scheduled
 * round count: an overtime bout ("3 Rnd + OT (5-5-5-5)") reports 4 periods, and an unfilled record reports 0. The
 * canonical convention, set by the historical record (42 of 43 "3 Rnd + OT" bouts), is scheduled_rounds = REGULATION
 * rounds, overtime not counted. Resolution, strongest evidence first:
 *   1. the structured time-format description ("N Rnd ..."), cross-checked against periods (N, or N + OT periods);
 *   2. periods alone, only when it is 3 or 5 (the only UFC regulation lengths);
 *   3. a title bout with no usable count: 5 (championship bouts are five rounds under the Unified Rules);
 * otherwise UNRESOLVED (rounds null). Never 0 or 4, and a non-title main event is never assumed to be five.
 * Returns { rounds, overtime, state: 'resolved'|'unresolved'|'conflict', basis, periods, description }. */
export function resolveScheduledRounds({ periods = null, description = null, isTitle = false } = {}) {
  const desc = String(description || '').trim();
  const p = Number.isInteger(periods) ? periods : null;
  const m = desc.match(/^\s*(\d+)\s*Rnd(?:\s*\+\s*(\d*)\s*OT)?/i);
  const out = (rounds, state, basis, overtime = 0) => ({ rounds, overtime, state, basis, periods: p, description: desc || null });
  if (m) {
    const reg = Number(m[1]);
    const ot = m[2] !== undefined ? (m[2] === '' ? 1 : Number(m[2])) : 0;
    const valid = reg >= 1 && reg <= 5;
    const periodsAgree = p === null || p === 0 || p === reg || p === reg + ot;
    if (!valid) return out(null, 'unresolved', 'description_out_of_range', ot);
    if (!periodsAgree) return out(null, 'conflict', 'description_vs_periods', ot);
    if (isTitle && reg !== 5) return out(null, 'conflict', 'title_bout_not_five_rounds', ot);
    return out(reg, 'resolved', ot ? 'description_regulation_plus_overtime' : 'description', ot);
  }
  if (p === 3 || p === 5) {
    if (isTitle && p !== 5) return out(null, 'conflict', 'title_bout_not_five_rounds');
    return out(p, 'resolved', 'periods');
  }
  if (isTitle) return out(5, 'resolved', 'title_bout_rule');
  return out(null, 'unresolved', p === null ? 'no_round_data' : `invalid_periods_${p}`);
}
